src/libraries/Common/src/System/Security/Cryptography/RSAAndroid.cs

```diff
@@ -422,47 +426,56 @@ public override void ImportParameters(RSAParameters parameters)
-                // with the already loaded key.
-                ForceSetKeySize(BitsPerByte * Interop.AndroidCrypto.RsaSize(key));
+                SetKeySizeFromHandle(key);
             }
 
-            public override void ImportRSAPublicKey(ReadOnlySpan<byte> source, out int bytesRead)
+            public override unsafe void ImportRSAPublicKey(ReadOnlySpan<byte> source, out int bytesRead)
             {
                 ThrowIfDisposed();
 
-                int read;
-
-                try
-                {
-                    AsnDecoder.ReadEncodedValue(
-                        source,
-                        AsnEncodingRules.BER,
-                        out _,
-                        out _,
-                        out read);
-                }
-                catch (AsnContentException e)
+                fixed (byte* ptr = &MemoryMarshal.GetReference(source))
                 {
-                    throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding, e);
-                }
-
-                SafeRsaHandle key = Interop.AndroidCrypto.DecodeRsaPublicKey(source.Slice(0, read));
-
-                if (key is null || key.IsInvalid)
-                {
-                    throw new CryptographicException();
-                }
+                    using (MemoryManager<byte> manager = new PointerMemoryManager<byte>(ptr, source.Length))
+                    {
+                        ReadOnlyMemory<byte> subjectPublicKey;
+                        try
+                        {
+                            AsnReader reader = new AsnReader(manager.Memory, AsnEncodingRules.BER);
+                            subjectPublicKey = reader.PeekEncodedValue();
+                        }
+                        catch (AsnContentException e)
+                        {
+                            throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding, e);
+                        }
 
-                FreeKey();
-                _key = new Lazy<SafeRsaHandle>(key);
+                        // Decoding the key on Android expects the encoded SubjectPublicKeyInfo,
+                        // not just the SubjectPublicKey, so we construct one.
+                        SubjectPublicKeyInfoAsn spki = new SubjectPublicKeyInfoAsn
+                        {
+                            Algorithm = new AlgorithmIdentifierAsn
+                            {
+                                Algorithm = Oids.Rsa,
+                                Parameters = AlgorithmIdentifierAsn.ExplicitDerNull,
+                            },
+                            SubjectPublicKey = subjectPublicKey,
+                        };
+
+                        AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
+                        spki.Encode(writer);
+
+                        SafeRsaHandle key = Interop.AndroidCrypto.DecodeRsaPublicKey(writer.Encode());
```

Comment:
If the function takes SPKI instead of RSAPublicKey (which is a data type), please rename it to DecodeRsaSubjectPublicKeyInfo, or something like that.